<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <!-- Define arguments for use in the launch file -->
  <arg name="scan_topic" default="scan" />
  <arg name="map_file" default="$(find bcr_bot)/maps/no_roof_small_warehouse.yaml" />

  <!-- Publish map topic -->
  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)" />

  <!-- Launch AMCL for each robot -->
  <group ns="bcr_bot_0">
    <arg name="robot_namespace" value="bcr_bot_0" />
    <arg name="initial_pose_x" default="0.0" />
    <arg name="initial_pose_y" default="0.0" />
    <arg name="initial_pose_a" default="0.0" />

    <node pkg="amcl" type="amcl" name="amcl">
      <param name="initial_pose_x" value="$(arg initial_pose_x)" />
      <param name="initial_pose_y" value="$(arg initial_pose_y)" />
      <param name="initial_pose_a" value="$(arg initial_pose_a)" />
      <param name="odom_frame_id" value="$(arg robot_namespace)/odom" />
      <param name="base_frame_id" value="$(arg robot_namespace)/base_link" />
      <param name="global_frame_id" value="map" />
      <remap from="scan" to="$(arg robot_namespace)/$(arg scan_topic)" />
      <remap from="initialpose" to="$(arg robot_namespace)/initialpose" />
      <remap from="amcl_pose" to="$(arg robot_namespace)/amcl_pose" />
      <remap from="particlecloud" to="$(arg robot_namespace)/particlecloud" />
    </node>
  </group>

  <group ns="bcr_bot_1">
    <arg name="robot_namespace" value="bcr_bot_1" />
    <arg name="initial_pose_x" default="0.0" />
    <arg name="initial_pose_y" default="2.0" />
    <arg name="initial_pose_a" default="0.0" />

    <node pkg="amcl" type="amcl" name="amcl">
      <param name="initial_pose_x" value="$(arg initial_pose_x)" />
      <param name="initial_pose_y" value="$(arg initial_pose_y)" />
      <param name="initial_pose_a" value="$(arg initial_pose_a)" />
      <param name="odom_frame_id" value="$(arg robot_namespace)/odom" />
      <param name="base_frame_id" value="$(arg robot_namespace)/base_link" />
      <param name="global_frame_id" value="map" />
      <remap from="scan" to="$(arg scan_topic)" />
      <remap from="initialpose" to="$(arg robot_namespace)/initialpose" />
      <remap from="amcl_pose" to="$(arg robot_namespace)/amcl_pose" />
      <remap from="particlecloud" to="$(arg robot_namespace)/particlecloud" />
    </node>
  </group>

  <group ns="bcr_bot_2">
    <arg name="robot_namespace" value="bcr_bot_2" />
    <arg name="initial_pose_x" default="-2.0" />
    <arg name="initial_pose_y" default="0.0" />
    <arg name="initial_pose_a" default="0.0" />

    <node pkg="amcl" type="amcl" name="amcl">
      <param name="initial_pose_x" value="$(arg initial_pose_x)" />
      <param name="initial_pose_y" value="$(arg initial_pose_y)" />
      <param name="initial_pose_a" value="$(arg initial_pose_a)" />
      <param name="odom_frame_id" value="$(arg robot_namespace)/odom" />
      <param name="base_frame_id" value="$(arg robot_namespace)/base_link" />
      <param name="global_frame_id" value="map" />
      <remap from="scan" to="$(arg scan_topic)" />
      <remap from="initialpose" to="$(arg robot_namespace)/initialpose" />
      <remap from="amcl_pose" to="$(arg robot_namespace)/amcl_pose" />
      <remap from="particlecloud" to="$(arg robot_namespace)/particlecloud" />
    </node>
  </group>

</launch>